### PR TITLE
Add Auto Deployments with CloudRun + GitHub Actions

### PR DESCRIPTION
**Pull request by SudharakaP**
This basically takes away most of the developer effort in deploying jhipster-online, and automates the deployment using CloudRun + GitHub Actions.

Fixes https://github.com/jhipster/jhipster-online/issues/172

### CODE REVIEW

**Comment by pascalgrimaud** [comment]
@SudharakaP : to understand well, what happened once this PR is merged ? Will it replace definitively the original deployment in GCE ? Should we stop GCE before ?

**Comment by SudharakaP** [comment]
@pascalgrimaud : Nothing will happen and it will not stop the GCE deployment since the GCE deployment is all manual, this PR has no effect on it. 

After we merge this PR, and create a release (with a tag say v2.6.2), then the deployment to CloudRun will trigger. Even then nothing will happen to the GCE deployment. The CloudRun deployment will be separate from the GCE deployment. 

We can then test if the CloudRun deployment is working perfectly and if so, we have to setup the custom domain in CloudRun (so that CloudRun points to https://start.jhipster.tech/). After that we can manually shutdown the GCE. 

Bottom line this PR merging will not automatically do anything to GCE. We have to manually shut it down after we do the testing and we are totally sure the CloudRun deployment works as expected. :smile:

**Comment by pascalgrimaud** [comment]
@SudharakaP : nice ! I though about https://start-beta.jhipster.tech/ or something else

**Comment by PierreBesson** [comment]
I think your solution is good @SudharakaP ! It's best to have secrets in one place like GitHub Secrets. However we must make sure that access control to the secrets value is safeguarded and that they can only be unencrypted from the master branch.

Maybe @jdubois can review the setup in terms of security because I'm not familiar with GitHub Secrets.

File: Dockerfile
Comment (vishal423) [inline Dockerfile:10]: If we are using this docker file for exclusive gcp deployments, then, I suggest renaming to reflect the context. Further, the maven `gcp` profile includes the spring `prod` profile.

**Comment by SudharakaP** [comment]
> I think your solution is good @SudharakaP ! It's best to have secrets in one place like GitHub Secrets. However we must make sure that access control to the secrets value is safeguarded and that they can only be unencrypted from the master branch.
> 
> Maybe @jdubois can review the setup in terms of security because I'm not familiar with GitHub Secrets.

@PierreBesson : Note that the deployment action is only triggered when creating a tag and therefore I believe only people in the core team can do this. Just by pushing things to branches won't trigger the deployment action. :smile:

File: src/main/resources/config/application-prod.yml
Comment (vishal423) [inline src/main/resources/config/application-prod.yml:29]: I suggest moving these to separate `gcp` specific profile

File: src/main/resources/config/application-prod.yml
Comment (vishal423) [inline src/main/resources/config/application-prod.yml:53]: same as above, production profile properties shouldn't be tied with `gcp`

File: src/main/resources/config/application-prod.yml
Comment (vishal423) [inline src/main/resources/config/application-prod.yml:64]: any specific reason to use `$SECRET_VALUE` in all places? I prefer to revert unless they serve a specific purpose

File: src/main/resources/config/application-prod.yml
Comment (vishal423) [inline src/main/resources/config/application-prod.yml:92]: you can always override with environment variables like `SERVER_PORT`

**Comment by SudharakaP** [comment]
> I think your solution is good @SudharakaP ! It's best to have secrets in one place like GitHub Secrets. However we must make sure that access control to the secrets value is safeguarded and that they can only be unencrypted from the master branch.
> 
> Maybe @jdubois can review the setup in terms of security because I'm not familiar with GitHub Secrets.

@PierreBesson : I've changed the deployment action to be only triggered when publishing a release as I think it's even better than relying on the tag. That means only people in the core team or people who has access to create releases can make the internal decryption happen. Even then they will not see the secrets since they are not recorded anywhere. :smile:

File: src/main/resources/config/application-prod.yml
Comment (SudharakaP) [reply src/main/resources/config/application-prod.yml:64]: Good point. I thought about it for a moment and put it there just so that we could keep track of which values are being replaced. However yes it can be removed. :smile:

File: src/main/resources/config/application-prod.yml
Comment (SudharakaP) [reply src/main/resources/config/application-prod.yml:92]: This PORT variable is GCP specific. It has nothing to do with a GiHub environment variable. :smile:

File: .github/workflows/deploy.yml
Comment (vishal423) [inline .github/workflows/deploy.yml]: It would be good to restrict this workflow to the `master` branch

Comment (SudharakaP) [reply .github/workflows/deploy.yml]: Thanks. Just did it a while ago after Pierre's comment above. I've restricted for the workflow to be triggered when publishing releases; which I think is better. :smile:

File: Dockerfile
Comment (SudharakaP) [reply Dockerfile]: You are correct; I've changed it to only include the gcp profile. As for renaming the docker file; currently that's not possible. Because, `gcloud builds submit` command expects the file to be named as `Dockerfile`. :cry:

File: src/main/resources/config/application-prod.yml
Comment (SudharakaP) [reply src/main/resources/config/application-prod.yml]: Thanks. Done. :+1:

File: src/main/resources/config/application-prod.yml
Comment (SudharakaP) [reply src/main/resources/config/application-prod.yml]: Thanks. Done. :+1:

File: src/main/resources/config/application-prod.yml
Comment (SudharakaP) [reply src/main/resources/config/application-prod.yml]: Done. :+1:

File: src/main/resources/config/application-gcp.yml
Comment (vishal423) [inline src/main/resources/config/application-gcp.yml:50]: unless these override a value provided in base application.yml or other profile-specific yml (prod/dev), I will prefer to not duplicate here.

File: src/main/resources/config/application-gcp.yml
Comment (vishal423) [inline src/main/resources/config/application-gcp.yml:30]: unless these override a value provided in base application.yml or other profile-specific yml (prod/dev), I will prefer to not duplicate here.

File: src/main/resources/config/application-gcp.yml
Comment (SudharakaP) [reply src/main/resources/config/application-gcp.yml]: Thanks :+1:

File: src/main/resources/config/application-gcp.yml
Comment (SudharakaP) [reply src/main/resources/config/application-gcp.yml:30]: Thanks. I've removed the duplicate properties that are present in `application-gcp.yml`. However properties such as `jdbcUrl` are not defined in `application-prod.yml` or `application.yml` so I think I have to keep them there for the substitutions. :smile: :thinking:

**Comment by SudharakaP** [comment]
@jdubois @pascalgrimaud : Just pinging to see if we could merge this; or if anything else needs to be done? This I think would make the deployments a lot simpler. :smile:

**Comment by pascalgrimaud** [comment]
@SudharakaP : I think you know better than me, what is the best here. But if we decide to use this new deployment, it's important to document it somewhere, in case you're in vacations for example. So we'll know how to fix when there is a problem
- how to deploy
- how to restart
- how to see log
- etc...
cc @PierreBesson too

**Comment by SudharakaP** [comment]
> @SudharakaP : I think you know better than me, what is the best here. But if we decide to use this new deployment, it's important to document it somewhere, in case you're in vacations for example. So we'll know how to fix when there is a problem
> 
>     * how to deploy
> 
>     * how to restart
> 
>     * how to see log
> 
>     * etc...
>       cc @PierreBesson too

@pascalgrimaud : I've just added some details about these things to the readme. If you have any other things that need to be added let me know. :smile: 

Also do we do a docker hub deployment in parallel when we do a new release? Is so I think I can add some code to automate that as well. :smile:

**Comment by SudharakaP** [comment]
@jdubois @PierreBesson : I've added the environment variables to the `app.yaml` file added a `.dockerignore` file so that the `app.yaml` file is ignored in the docker image. I hope this solves the docker image containing secrets problem. Also I've fixed the email problem we saw. Give it a try and let me know if anything else needs to be done. 😄

**Review by PierreBesson — APPROVED** [review]

**Comment by PierreBesson** [comment]
@SudharakaP It looks good to me. Let's start the migration !

**Comment by SudharakaP** [comment]
@PierreBesson : Okay, so shall I merge this PR and try to add the custom domain. Or maybe we can do it together over a call?

**Comment by PierreBesson** [comment]
I trust you to do it. If you need anything just ping me.

**Comment by saturnism** [comment]
Congrats! Glad to see the migration was successful :D

**Comment by SudharakaP** [comment]
@saturnism : Thanks much for your help on this. Greatly appreciated. 👍🏼

**Pull request merged**
